zerver/lib/widget.py

```diff
@@ -46,6 +48,28 @@ def do_widget_post_save_actions(message: MutableMapping[str, Any]) -> None:
     if content in ['/poll', '/tictactoe']:
         widget_type = content[1:]
 
+    widget_content = message.get('widget_content')
+    if widget_content is not None:
+        # We cannot trust any incoming content, so we catch
+        # all exceptions and early-exit.
+        try:
+            widget_content = json.loads(widget_content)
+        except Exception:
+            logging.warning('API user sent invalid content')
```

Comment:
I think we want this validation to happen inside `check_message` and return a `JsonableError` so that the API client gets the error, not the server logs.

Reply:
(And then once that parsing is happily done, it can just store on the proto-message object the happily parsed and validated `extra_data`)